dhall-nix/src/Dhall/Nix.hs

```diff
@@ -738,3 +728,129 @@ dhallToNix e =
     loop (ImportAlt a _) = loop a
     loop (Note _ b) = loop b
     loop (Embed x) = absurd x
+
+-- | Previously we turned @<Foo | Bar>.Foo@ into @{ Foo, Bar }: Foo@,
+-- but this would not work with <Frob/Baz>.Frob/Baz (cause the slash is not a valid symbol char in nix)
+-- so we generate @union: union."Frob/Baz"@ instead.
+--
+-- If passArgument is @Just@, pass the argument to the union selector.
+unionChoice :: Text -> Maybe NExpr -> NExpr
+unionChoice chosenKey passArgument =
+  let selector = Fix (Nix.NSelect Nothing (Nix.mkSym "u") (mkDoubleQuotedIfNecessary chosenKey :| []))
+  in Nix.Param "u" ==>
+     case passArgument of
+       Nothing -> selector
+       Just arg -> selector @@ arg
+
+
+-- | Double-quote a field name (record or union). This makes sure it’s recognized as a valid name by nix, e.g. in
+--
+--  @{ "foo/bar" = 42; }."foo/bar" }@
+--
+-- where
+--
+-- @{ foo/bar = 42; }.foo/bar@ is not syntactically valid nix.
+--
+-- This is only done if necessary (where “necessary” is not super defined right now).
+mkDoubleQuotedIfNecessary :: Text -> NKeyName r
+mkDoubleQuotedIfNecessary key =
+    if Text.all simpleChar key
+    then StaticKey (VarName key)
+    else DynamicKey (Plain (DoubleQuoted [Plain key]))
+    where
+        simpleChar c = isAsciiLower c || isAsciiUpper c
```

Comment:
Minor suggestion: I think this makes the intention more clear

```suggestion
        simpleChar c = isAscii c || isLetter c
```

This would also require a matching change to the imports

Reply:
I … don’t think all ascii values are a valid nix symbol. It includes stuff like 0x04 (Ctrl+c); and I’d have to verify if they are allowed to start with a digit

Reply:
For example

```
nix-repl> { 1 = "foo"; }
error: syntax error, unexpected INT

       at «string»:1:3:

            1| { 1 = "foo"; }
             |   ^
            2|
```

Reply:
Whoops, I meant to say `isAscii c && isLetter c`

Reply:
But `isAscii` still includes all control characters for example

Reply:
The additional `isLetter` requirement would exclude those control characters

Reply:
i.e. 

```
-- | Selects the first 128 characters of the Unicode character set,
-- corresponding to the ASCII character set.
isAscii :: Char -> Bool
isAscii c = c < '\x80'
```

Reply:
Please read my responses more closely

Reply:
Oh, using `&&` … at least to me filtering out by boolean and is not more intuitive than set logic.

Reply:
The main reason I suggested using `isLetter` is that it seems indirect to specify that something is a letter by saying that it is uppercase or lowercase.  What you've written reads to me like:

```haskell
simpleChar c = (isAscii c && isUpper c) || (isAscii c && isLower c)
```

Reply:
Anyway, coming back to this, apart from the aesthetics, `isLetter` needs support for unicode ICU (it’s not just a simple `ord` check):

```
isLetter :: Char -> Bool
isLetter c = case generalCategory c of
 GeneralCategory
 UppercaseLetter -> True
 LowercaseLetter -> True
 TitlecaseLetter -> True
 ModifierLetter -> True
 OtherLetter -> True
 _ -> False

generalCategory :: Char -> GeneralCategory
generalCategory c = toEnum $ fromIntegral $ wgencat $ fromIntegral $ ord c

foreign import ccall unsafe "u_gencat"
 wgencat :: Int -> Int
```

vs 

```
isAsciiLower :: Char -> Bool
isAsciiLower c = c >= 'a' && c <= 'z'
```